feat: export maven/gradle project

tadayosi opened the pull request:
My initial attempt to fix #1044. I know there are still rough edges but at least the following basic usages work. Let me send it for your review so that I can check I'm on the right track.

Add two subcommands to 'jbang export':
* jbang export maven
* jbang export gradle
 
Those generate a template maven/gradle project the user can start developing the script further.

Fix #1044

Usage examples:
```console
jbang export maven hello.java # generates hello/ project with org.example.project:hello:999-SNAPSHOT
jbang export gradle -g dev.jbang.hello -a hello -v 1.0.0 -O hello2 hello.java # generates hello2/ with dev.jbang.hello:hello:1.0.0
```

tadayosi commented:
@maxandersen @quintesse Please review this.

quintesse commented on src/main/java/dev/jbang/cli/Export.java:
```diff
@@ -502,3 +498,234 @@ private Path getJlinkOutputPath() {
+	private void createProjectForExport(BuildContext ctx, Path projectDir) throws IOException {
+		Project prj = ctx.getProject();
+		String projectName = projectDir.getFileName().toString();
+		Path tmpProjectDir = Files.createTempDirectory(projectName);
```
This line, plus the `Files.move()` on line 590 won't work on Windows if the directory pointed to by `projectDir` is on a different drive than `%TEMP%`.

So I'd just skip the temp folder and immediately copy/create any files directly into the target folder.

tadayosi replied:
Thanks! I will change it accordingly.

quintesse reviewed (changes requested):
Looks good! Fix the one thing I mentioned and I think this should be ready to merge.

tadayosi commented:
@quintesse Applied your feedback. Thanks!

quintesse reviewed (dismissed):
Code-wise it all looks good to me! Perhaps you could add a couple of tests?

quintesse commented:
@tadayosi please be advised I updated your PR. I rebased it on top of main to have it include some recent fixes to the CI process. Without that the CI build would fail.

tadayosi commented:
@quintesse Added tests for basic scenarios.

tadayosi commented:
@quintesse Could you tell me if there's anything else that blocks it from merging?  Thank you!

quintesse reviewed (dismissed):
LGTM

quintesse commented:
> @quintesse Could you tell me if there's anything else that blocks it from merging? Thank you!

No, everything looks good. Now we just wait for @maxandersen to merge it. (But he's been very busy so it might take a bit of time)

maxandersen commented:
looking at this - looks great thus far but really should have something in docs on this. wanna add that in this pr ?

maxandersen commented on src/main/java/dev/jbang/cli/Export.java:
```diff
@@ -502,3 +498,239 @@ private Path getJlinkOutputPath() {
+		Path destination = projectDir.resolve("pom.xml");
+
+		Map<String, String> properties = new HashMap<>();
+		if (prj.getJavaVersion() != null) {
```
default jbang java is 8...should we explicilty set it?

tadayosi replied:
While it totally makes sense for building JBang itself, I'd guess Java 8 is a bit too restrictive for normal JBang users. I'd imagine most users just expect the default local Java version to be used unless they explicitly set `//JAVA`. I think this can be achieve just by using `JavaUtil.javaVersion`. Let me update the code accordingly and make it always set a Java version.

tadayosi replied:
Ah, thanks for your explanation. So, I agree that we should only set the version if the script explicitly set `//JAVA`. As Tako said, it seems to be really an edge case where the user fails with it.

I'll go ahead and change the PR again so that it strictly follows `//JAVA` and when absent it won't set the Java versions.

maxandersen replied:
i'm -0 on it but lets just see - we can change it later if problematic :)

maxandersen commented on src/main/java/dev/jbang/cli/Export.java:
```diff
@@ -502,3 +498,239 @@ private Path getJlinkOutputPath() {
+
+		Map<String, String> properties = new HashMap<>();
+		if (prj.getJavaVersion() != null) {
+			properties.put("maven.compiler.source", prj.getJavaVersion());
```
this is not handling case of `//JAVA 17+` it shuold remove the +.

tadayosi replied:
Thanks for pointing it out. I'll modify it to use `JavaUtil.javaVersion` which should handle `+` cases.

tadayosi replied:
Done

maxandersen commented:
this could probably be handled in updates but just listing here for awareness:

ignores //PREVIEW, breaks on kotlin/other languages, quarkus integration not handled (so `jbang -t qli test.java; jbang export maven test.java` doesnt work), and probably more :)

but definitely a win.

lets at least fix the java version and some basic docs about it.

tadayosi commented:
@maxandersen Thanks for your review! Fixed the java version issue and added basic docs for the new subcommand to `exporting.adoc`. If I'm not mistaken, the CLI Reference docs are something auto-generated later, is that right?  Please review this again.

tadayosi commented:
@maxandersen @quintesse Fixed the java version issue again by applying Tako's suggestion. Now I believe using `JavaUtil.minRequestedVersion()` is the right way. I hope this finally gets it right.

maxandersen commented on docs/modules/ROOT/pages/exporting.adoc:
```diff
@@ -45,3 +47,110 @@ You should only need to change `hello.java` to match your application/script.
+
+[source,bash]
+----
+jbang export gradle --group dev.jbang --artifact myapp --version 1.0.0-SNAPSHOT hello.java
```
```suggestion
jbang export gradle --group org.acme --artifact myapp --version 1.0.0-SNAPSHOT hello.java
```

lets not confuse use of the jbang namespace.

tadayosi replied:
Done

maxandersen commented:
@all-contributors please add @tadayosi for code

tadayosi commented:
@maxandersen @quintesse Thanks!

maxandersen commented:
just realized this is moving default package names into package names based on GAV - that isn't right. working on a patch.
